### PR TITLE
hermes-mock | Fix matching avro messages with nullable fields

### PR DESCRIPTION
**Pull request by sobelek**
Hey!

During our development we noticed a problem in hermes-mock.  If there exists stubs for 2 different topics where one of them has some set of nullable fields it is possible that with current implementation parsing message to given schema can produce unhandled exception. And It can prevent hermes-mock from matching correct message to stub.
Added tests to confirm the issue.

### CODE REVIEW

**Review by szczygiel-m — APPROVED** [review]

File: hermes-mock/src/test/groovy/pl/allegro/tech/hermes/mock/HermesMockAvroTest.groovy [hunk 1]
```diff
@@ -71,6 +72,39 @@ class HermesMockAvroTest extends Specification {
...
+            def response2 = publish(topicName2, message2)
+
+        then: "check for any single message on the topics and check for correct responses"
+            hermes.expect().singleMessageOnTopic(topicName)
+            hermes.expect().singleMessageOnTopic(topicName2)
+
+            response.status == HttpStatus.SC_CREATED
+            response2.status == HttpStatus.SC_CREATED
+    }
+
```
Comment (pitagoras3) [inline hermes-mock/src/test/groovy/pl/allegro/tech/hermes/mock/HermesMockAvroTest.groovy:107]: **nitpick:** empty blank line

You can remove it to match file formatting style (1 empty line between sections).

**Review by pitagoras3 — APPROVED** [review]

File: hermes-mock/src/test/groovy/pl/allegro/tech/hermes/mock/HermesMockAvroTest.groovy [hunk 2]
```diff
@@ -71,6 +72,39 @@ class HermesMockAvroTest extends Specification {
             response2.status == HttpStatus.SC_CREATED
     }
 
+    def "should receive an Avro message matched by pattern different nullable schema"() {
+        given: "define wiremock responses for 2 topics with different schemas"
+            def topicName = "my-test-avro-topic-1"
+            def topicName2 = "my-test-avro-topic-2"
+
```
[the remark this answers is not in the document]
Comment (sobelek) [reply hermes-mock/src/test/groovy/pl/allegro/tech/hermes/mock/HermesMockAvroTest.groovy:79]: done

File: hermes-mock/src/test/groovy/pl/allegro/tech/hermes/mock/HermesMockAvroTest.groovy (hunk 1 above)
Comment (sobelek) [reply hermes-mock/src/test/groovy/pl/allegro/tech/hermes/mock/HermesMockAvroTest.groovy:105]: done

File: hermes-mock/src/test/groovy/pl/allegro/tech/hermes/mock/HermesMockAvroTest.groovy [hunk 3]
```diff
@@ -378,6 +415,9 @@ class HermesMockAvroTest extends Specification {
     def publish(String topic, TestMessageWithDifferentSchema message) {
         publish(topic, asAvro(message))
     }
+    def publish(String topic, TestMessageWithNullableSchema message) {
```
[the remark this answers is not in the document]
Comment (sobelek) [reply hermes-mock/src/test/groovy/pl/allegro/tech/hermes/mock/HermesMockAvroTest.groovy:417]: done

File: hermes-mock/src/test/groovy/pl/allegro/tech/hermes/mock/TestMessage.groovy [hunk 4]
```diff
@@ -41,3 +41,11 @@ class TestMessageWithDifferentSchema {
         JsonOutput.toJson(this)
     }
 }
+
+class TestMessageWithNullableSchema extends TestMessageWithDifferentSchema{
```
[the remark this answers is not in the document]
Comment (sobelek) [reply hermes-mock/src/test/groovy/pl/allegro/tech/hermes/mock/TestMessage.groovy]: done

**Comment by sobelek** [comment]
@pitagoras3 You are right. I was thinking about catching only `ArrayIndexOutOfBoundsException` but ultimately decided to catch any runtime one. After all I would like hermes-mock to just not match request instead of crashing.

**Review by szczygiel-m — APPROVED** [review]

**Pull request merged**